Adding client side report generation

jacob6838 opened the pull request:
<!--- Thanks for the contribution! -->

# PR Details

## Description

Adding Intersection Client Side Report Generation
**The changes in this PR were copied directly from https://github.com/CDOT-CV/jpo-cvmanager/pull/138**

This PR adds client-side report generation to the jpo-cvmanager. This feature was [originally developed](https://github.com/usdot-jpo-ode/jpo-conflictvisualizer/pull/130/files) for the conflictvisualizer, and was carried into the jpo-cvmanager to be compatible with changes made to the intersection api.

Data for reports is aggregated in the background, and stored in MongoDB. When a report is pre-viewed, the API provides the data from MongoDB, and the webapp renders the report client side. When the report is download, the webapp uses jspdf to convert the rendered report into a pdf. 

Ex: 
Client-side Preview:
![image](https://github.com/user-attachments/assets/eb84b200-8703-4a4f-b82a-b7e0a656bd1e)

Downloading:
![image](https://github.com/user-attachments/assets/6cd08c93-fb17-400a-a1d4-59eb85aec5e6)

PDF: 
![image](https://github.com/user-attachments/assets/15f19f98-e37d-42e1-956f-ab1cc7e1da4a)



## How Has This Been Tested?

These changes were tested by running the webapp alongside the intersection api. To generate a report, go to the intersectionDashboard/reports page, select the intersection '12109', then select "Generate Manual Report", and include at a minimum the day 2024-08-20 (this is when most of the pre-loaded mocked data is from)

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If a section applies, ensure you have met all of the associated checks -->

- [ ] My changes require new environment variables:
  - [ ] I have updated the docker-compose, K8s YAML, and all dependent deployment configuration files.
- [ ] My changes require updates to the documentation:
  - [ ] I have updated the documentation accordingly.
- [ ] My changes require updates and/or additions to the unit tests:
  - [ ] I have modified/added tests to cover my changes.
- [x] All existing tests pass.

jacob6838 committed:
Adding client side report generation

payneBrandon commented on webapp/src/apis/intersections/reports-api.ts:
```diff
@@ -8,6 +13,38 @@ export type ReportMetadata = {
   reportStartTime: Date
   reportStopTime: Date
   reportContents: string[]
+  laneDirectionOfTravelEventCounts: { id: string; count: number }[]
```
Would it make sense to abstract the id/count to an object since we use the same structure so often?

jacob6838 replied:
Created CountWithId type

payneBrandon commented on webapp/src/components/intersections/ReportsPage.tsx:
```diff
@@ -129,6 +129,28 @@ const Page = () => {
+    setOpenReportGenerationDialog(false)
+    const refreshTime = new Date(Date.now() + 15 * 60 * 1000)
+    toast.success(
+      `Reports usually take 10-15 minutes to generate. Please refresh the page at ${refreshTime.toLocaleTimeString()} to see the new report.`
```
We spoke offline on this a little at least with the data generation portion, but is there any notification that the report is ready for download? Or is it solely on the user to come back in 15 minutes to check?

jacob6838 replied:
Right now, just on the user to check. The report generation is asynchronous

payneBrandon commented on webapp/src/features/intersections/reports/report-list-table.tsx:
```diff
@@ -1,56 +1,18 @@
-  TableRow,
-  Typography,
-} from '@mui/material'
+import { Box, Button, Table, TableBody, TableCell, TablePagination, TableRow, Typography } from '@mui/material'
```
For these horizontal vs vertical include list, is that a setting we can set to keep consistent?

jacob6838 replied:
Right now, it condenses if it all fits onto one line (length 120), and expands to new lines if it doesn't. It looks like we can also add eslint, which will expand every import onto multiple lines - is that a better solution?

payneBrandon replied:
Sticking with one or the other is preferable, it's strange to see the change in a file when it's only this formatting. Makes it difficult to determine what to review

payneBrandon commented on webapp/src/features/intersections/reports/report-utils.ts:
```diff
@@ -0,0 +1,118 @@
+  return readableStrings
+}
+
+export const formatNumber = (num: number) => {
```
Should we be more explicit in the naming here or add a comment so we know what format does?

jacob6838 replied:
re-named to formatAxisTickNumber

payneBrandon reviewed (changes requested):
A few comments

jacob6838 committed:
Creating CountWithId type

jacob6838 committed:
Re-name formatNumber to formatAxisTickNumber

jacob6838 committed:
Fixing report refresh

mcook42 commented on webapp/src/components/intersections/ReportsPage.tsx:
```diff
@@ -166,6 +189,22 @@ const Page = () => {
+                  onClick={() =>
+                    setFilters({
+                      ...filters,
+                      startDate: new Date(new Date().getTime() - 7 * 24 * 60 * 60 * 1000),
```
sugestion(non-blocking): define a named const for `7 * 24 * 60 * 60 * 1000` so that it's clear what this calculation represents && we don't have to recalculate this value on each re-render (it can be reused each time we do the calculation like `new Date(new Date().getTime() - <named const>)`)

jacob6838 replied:
updated

payneBrandon replied:
Looks like this instance is still using the calculation. I see others were updated though

jacob6838 replied:
updated

mcook42 commented on webapp/src/features/intersections/reports/graphs/bar-chart-component.tsx:
```diff
@@ -0,0 +1,63 @@
+}
+
+const CustomTooltip = ({ active, payload, label }: TooltipProps<ValueType, NameType>) => {
+  if (active && payload && payload.length) {
```
suggestion(non-blocking): use optional chaining to reduce verbosity and improve readability
```suggestion
  if (active && payload?.length) {
```

jacob6838 replied:
updated

mcook42 commented on webapp/src/features/intersections/reports/report-color-palette.ts:
```diff
@@ -0,0 +1,14 @@
+const reportColorPalette = [
```
suggestion(non-blocking): could we use a map/dictionary instead of an array to improve readability? It's cumbersome (especially to a color-deficient guy like me) to have to use indexes to lookup the color in the array. I can do it, but using a map with named keys like the following could greatly improve the life of future maintainers

```ts
const reportColorPallettes = {
  'pink': '#C1666B',
  'orange-red': '#9a6d32',
  'yellow': '#d4b483',
  'grey': '#b5b1b2',
  'blue-grey': '#8c8698',
  'purple': '#745c97',
  'teal': '#48a9a6',
  'green': '#568259',
  'blue': '#4357ad',
  'cyan': '#98c1d9',
}
```

to access the colors you would do something like `reportColorPallettes['orange-red']` instead of `reportColorPallette[2]`

jacob6838 replied:
done

mcook42 commented on webapp/src/features/intersections/reports/report-list-table.tsx:
```diff
@@ -157,7 +114,7 @@ export const ReportListTable = (props: ReportListTableProps) => {
             <TableBody>
               {reports.map((report: ReportMetadata) => (
-                <ReportRow report={report} />
+                <ReportRow report={report} onViewReport={onViewReport} />
```
suggestion(non-blocking): define a key for the ReportRow. 
```suggestion
                <ReportRow key={report.reportName} report={report} onViewReport={onViewReport} />
```

From [Sonarqube](https://rules.sonarsource.com/typescript/impact/maintainability/RSPEC-6477/): 

> To optimize the rendering of React list components, a unique identifier (UID) is required for each list item. This UID lets React identify the item throughout its lifetime. To provide it, use the key attribute of the list item. When the key attribute is missing, React will default to using the item’s index inside the list component. If the element ordering changes, it will cause keys to not match up between renders, recreating the DOM. It can negatively impact performance and may cause issues with the component state.

mcook42 reviewed (commented):
some initial suggestions while I continue with a full review

mcook42 commented on webapp/src/features/intersections/reports/report-utils.ts:
```diff
@@ -0,0 +1,118 @@
+import { eachDayOfInterval, format, subDays } from 'date-fns'
+
+export type LaneDirectionOfTravelReportData = {
```
suggestion(non-blocking): We seem to frequently iterate over this data type by the `laneId` property (see [DistanceFromCenterlineGraphSet](https://github.com/CDOT-CV/jpo-cvmanager/blob/60d68f1e55742e8456b4cb0d3ba8c82984f4514e/webapp/src/features/intersections/reports/graphs/distance-from-centerline-graph-set.tsx#L16) and [HeadingErrorGraphSet](https://github.com/CDOT-CV/jpo-cvmanager/blob/60d68f1e55742e8456b4cb0d3ba8c82984f4514e/webapp/src/features/intersections/reports/graphs/heading-error-graph-set.tsx#L15)). It may be more efficient, be less error-prone, and reduce code duplication if we defined the type like so: 
```typescript
export type LaneDirectionOfTravelReportByLaneId = {
  [laneId: number]: LaneDirectionOfTravelReportData[]
}
```

An example of iterating over the datatype to get a collection of laneIds (like the code does in[ pdf-generator line 144](https://github.com/CDOT-CV/jpo-cvmanager/blob/60d68f1e55742e8456b4cb0d3ba8c82984f4514e/webapp/src/features/intersections/reports/pdf-generator.tsx#L144)) is:
```typescript
const laneIds = Object.keys(report).map(laneId => Number(laneId));
```

So instead of needing to duplicate the `groupedData` function seen in a few places like https://github.com/CDOT-CV/jpo-cvmanager/blob/60d68f1e55742e8456b4cb0d3ba8c82984f4514e/webapp/src/features/intersections/reports/graphs/distance-from-centerline-graph-set.tsx#L16 you could operate on the data type directly without needing to manipulate it first.

An example of simple usage in [DistanceFromCenterlineGraphSet](https://github.com/CDOT-CV/jpo-cvmanager/blob/60d68f1e55742e8456b4cb0d3ba8c82984f4514e/webapp/src/features/intersections/reports/graphs/distance-from-centerline-graph-set.tsx):

```typescript
interface DistanceFromCenterlineGraphSetProps {
    data: LaneDirectionOfTravelReportDataByLaneId
    distanceTolerance: number // New prop
}

const DistanceFromCenterlineGraphSet: React.FC<DistanceFromCenterlineGraphSetProps> = ({data, distanceTolerance}) => {
    return (
        <Box>
            {Object.keys(data).length === 0 ? (
                <Typography variant="body1" align="center">
                    No Data
                </Typography>
            ) : (
                Object.entries(data).map(
                    ([laneId, laneData]) => (
                        <Box key={laneId} id={`distance-from-centerline-graph-${laneId}`} sx={{mb: 6}}>
                            <DistanceFromCenterlineOverTimeGraph
                                data={laneData}
                                laneNumber={laneId.toString()}
                                distanceTolerance={distanceTolerance}
                            />
                        </Box>
                    )
                )
            )}
        </Box>
    )
}
```

Note: I didn't spend a ton of time digging into how best to implement the above examples, so there is likely a better way to do it, but at least this can show you how updating the data type definition can save you a ton of duplication and effort in the long run

jacob6838 replied:
Good point - I created this data type in report-details-modal using useMemo, then realized that nothing uses the raw type, only the map type, so just converted everything to the mapped type. Then made the same updates you recommended above - only had 1 duplication (2 total), but still nice to remove

mcook42 commented on webapp/src/features/intersections/reports/graphs/map-minimum-data-graph.tsx:
```diff
@@ -0,0 +1,19 @@
+
+const MapMinimumDataGraph: React.FC<MapMinimumDataGraphProps> = ({ data, getInterval }) => (
+  <BarChartComponent
+    title="MAP Minimum Data Events per Day"
```
Question(non-blocking): the only difference between this component definition and the definition of `webapp/src/features/intersections/reports/graphs/map-broadcast-rate-graph.tsx` above is the title. Is there a better approach that reduces code duplication while keeping implementations decoupled?

jacob6838 replied:
I removed all of the single-use and small graph react components, and put the <BarChartComponent> directly into the ReportDetailsModal component

mcook42 commented on webapp/src/features/intersections/reports/graphs/stop-line-stop-graph.tsx:
suggestion(non-blocking): delete unused code

jacob6838 replied:
removed

jacob6838 committed:
reportColorPallettes converted to named object

jacob6838 committed:
Removing single-use reports/graph components

jacob6838 committed:
Re-working laneDirectionOfTravelReportData to map

jacob6838 committed:
Update webapp/src/features/intersections/reports/report-list-table.tsx

Co-authored-by: Matt Cook <mattheworion.cook@gmail.com>

drewjj reviewed (approved):
Seems to be the same PR I approved previously. Approving again.

payneBrandon reviewed (changes requested):
just a couple questions

jacob6838 committed:
Update ReportsPage.tsx

jacob6838 committed:
Cleaning up unused variables/comments

jacob6838 committed:
Merge branch 'develop' into client-side-intersection-reports

jacob6838 committed:
Updating intersection-api local run command

jacob6838 committed:
Merge branch 'develop' into client-side-intersection-reports

jacob6838 committed:
Merge branch 'develop' into client-side-intersection-reports

payneBrandon commented:
@jacob6838  Do we have any unit tests for this new functionality? Or a user story in place to add them? It looks like we're pretty lacking in that department.

payneBrandon commented on webapp/src/components/intersections/ReportsPage.tsx:
```diff
@@ -94,17 +99,14 @@ const Page = () => {
 
   useEffect(
     () => {
+      console.log(filters)
```
leftover from testing?

jacob6838 replied:
removed

payneBrandon commented on webapp/src/components/intersections/ReportsPage.tsx:
```diff
@@ -177,13 +217,14 @@ const Page = () => {
           </Box>
           <ReportListTable
-            group={group}
+            group={true}
```
question: Should there be a variable we're connecting this property to? If not, do we need to open it as an option?

jacob6838 replied:
This was unused inside the table object - removed!

payneBrandon commented on webapp/src/features/intersections/reports/pdf-generator.tsx:
```diff
@@ -0,0 +1,556 @@
+  return currentPage
+}
+
+export const generatePdf = async (
```
suggestion: commenting/documentation on what this method is doing and what various components are setting. As it stands, this would be quite difficult for someone not very familiar with it to make any updates.

jacob6838 replied:
Just added a jsdoc and inline comments to this method

payneBrandon commented on .vscode/settings.json:
```diff
@@ -52,6 +52,7 @@
     "keyfile",
     "ksession",
     "loadercss",
+    "ldot",
```
question(non-blocking): what is ldot?

jacob6838 replied:
lane direction of travel. This was used in variables like "ldotReportData"

payneBrandon reviewed (changes requested):
A few questions

jacob6838 committed:
Cleaning up report table and logging

jacob6838 committed:
Documenting generatePdf method

jacob6838 committed:
Adding report-utils unit tests

jacob6838 committed:
Adding pdf-generator unit tests

jacob6838 committed:
Adding client side report webapp unit tests

jacob6838 committed:
Updating date-fns library version

jacob6838 committed:
Finalizing report-list-table unit tests

jacob6838 committed:
Mocking date-fns parse in unit tests

jacob6838 committed:
Updating report-list-table tests

payneBrandon commented:
@jacob6838 looks like a merge conflict with the other PR that was just pulled in

jacob6838 committed:
Merge branch 'develop' into client-side-intersection-reports

jacob6838 commented:
> @jacob6838 looks like a merge conflict with the other PR that was just pulled in

These are resolved!

rgelmis commented:
Some of the charts on the report are missing labelled axes, and this hover effect blocks the chart lines and covers the y-axis:
![image](https://github.com/user-attachments/assets/75eaf1dd-9c22-4fda-bb30-d1c52a2e15ae)
![image](https://github.com/user-attachments/assets/f09067e7-6923-46fb-a222-1373cc183068)
![image](https://github.com/user-attachments/assets/9c6354fc-a083-4bbc-861e-59fee264209f)

The yellow used here is very light and hard to see 
<img width="421" alt="image" src="https://github.com/user-attachments/assets/3d35b1e4-c322-4489-a1fc-2b98ac9c0a1f" />

payneBrandon commented:
@jacob6838 I'm having issues running the api here, let's touch base next week to resolve. I'd also like this to be able to run on a mac so our UI/UX folks can review and our devs using mac can contribute as well.

jacob6838 commented:
> Some of the charts on the report are missing labelled axes, and this hover effect blocks the chart lines and covers the y-axis: ![image](https://private-user-images.githubusercontent.com/111303732/434882663-75eaf1dd-9c22-4fda-bb30-d1c52a2e15ae.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUzNjcxMTYsIm5iZiI6MTc0NTM2NjgxNiwicGF0aCI6Ii8xMTEzMDM3MzIvNDM0ODgyNjYzLTc1ZWFmMWRkLTljMjItNGZkYS1iYjMwLWQxYzUyYTJlMTVhZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyM1QwMDA2NTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02Y2Y4YmY2ZTZiNzEwNmY2ZWY4MzYxYTYzODk1NTkxNTA4ZjNlMzQwNWZhNzM3NGM0YzM2MDkzNzgyYWIyYTI2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.pHVBxuTVeI5DVEBAs2zfiqdm9F71sZhZSEOU8TQ1NGY) ![image](https://private-user-images.githubusercontent.com/111303732/434882710-f09067e7-6923-46fb-a222-1373cc183068.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUzNjcxMTYsIm5iZiI6MTc0NTM2NjgxNiwicGF0aCI6Ii8xMTEzMDM3MzIvNDM0ODgyNzEwLWYwOTA2N2U3LTY5MjMtNDZmYi1hMjIyLTEzNzNjYzE4MzA2OC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyM1QwMDA2NTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xMzBiYTdkOWY4YzhkZmYyYTZiOTJiZTRhZDk4NjliMTg3ZDczNmM3YTNlNmVhNmMwYzViYjYxNzk1ZTdkNmNmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.QUa5Xj3R4_6E78iRPIOjldeGmKTqI6czj5_rg_Y7fG4) ![image](https://private-user-images.githubusercontent.com/111303732/434882601-9c6354fc-a083-4bbc-861e-59fee264209f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUzNjcxMTYsIm5iZiI6MTc0NTM2NjgxNiwicGF0aCI6Ii8xMTEzMDM3MzIvNDM0ODgyNjAxLTljNjM1NGZjLWEwODMtNGJiYy04NjFlLTU5ZmVlMjY0MjA5Zi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyM1QwMDA2NTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iNDI2ZjNhZGM2NzBjZjYzNTI3ZDNlZWE5MjZiMzAxN2I4ZTdlYmNhYjBmMDg3YjkyM2QyYWU5MTkxOWZmNWVhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.U4jEMkLR5k2nsVWIpCIdB9J4G8FaNHVNSQLmER_E0P0)
> 
> The yellow used here is very light and hard to see <img alt="image" width="421" src="https://private-user-images.githubusercontent.com/111303732/434883110-3d35b1e4-c322-4489-a1fc-2b98ac9c0a1f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUzNjcxMTYsIm5iZiI6MTc0NTM2NjgxNiwicGF0aCI6Ii8xMTEzMDM3MzIvNDM0ODgzMTEwLTNkMzViMWU0LWMzMjItNDQ4OS1hMWZjLTJiOThhYzljMGExZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQyM1QwMDA2NTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01NjYwYWQ5OTYxM2JlNmZjNmUyNTJjM2YzMjZjNzRlMzRkYmViNTJmZDY3MGMyNjAwYmFmYWFiNGE5N2U0MDE2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.DM0DQoUIBp0JweFa-KPKFDjwHMqWtaeH8QXxBcjyt9E">

I just updated the plots to only display when there is data available, that fixed the overflowing tooltip/hover bugs. I also changed the yellow color to be stronger. Here is a pic of both (no data, and stronger yellow):

![image](https://github.com/user-attachments/assets/8c729117-d057-4738-b156-86faf3bff37b)

jacob6838 committed:
Not displaying report graphs whe no data, changing report yellow color

jacob6838 committed:
Labeling report gen form and adding error messages

payneBrandon reviewed (changes requested):
It looks like the close button disappears when you scroll down through the popup, should this continue to be visible so users can close out at any point? Note that clicking outside the popup also doesn't close it

payneBrandon commented on webapp/src/components/intersections/ReportsPage.tsx:
```diff
@@ -9,8 +9,13 @@ import { ReportGenerationDialog } from '../../features/intersections/reports/rep
 import { selectToken } from '../../generalSlices/userSlice'
 import { useSelector } from 'react-redux'
+import ReportDetailsModal from '../../features/intersections/reports/report-details-modal'
+import toast from 'react-hot-toast'
```
suggestion: remove unused import

jacob6838 replied:
removed

payneBrandon commented on webapp/src/features/intersections/reports/report-list-table.tsx:
```diff
@@ -1,56 +1,18 @@
 }
 
 const ReportRow = (props: ReportRowProps) => {
   const navigate = useNavigate()
```
question: this doesn't seem to be used?

jacob6838 replied:
removed

payneBrandon reviewed (approved):
lgtm!

jacob6838 committed:
Converting report modal to dialog

jacob6838 commented:
> It looks like the close button disappears when you scroll down through the popup, should this continue to be visible so users can close out at any point? Note that clicking outside the popup also doesn't close it

Fixed both of these - I converted the report display to a dialog and put all of the controls at the bottom

rgelmis commented:
Just a general question on this functionality since I am able to run it now (yay!)- what is the difference between selecting a date under "Issue date" vs generate manual report > generate performance report > start date/end date?

jacob6838 commented:
> Just a general question on this functionality since I am able to run it now (yay!)- what is the difference between selecting a date under "Issue date" vs generate manual report > generate performance report > start date/end date?

The issued date is searching for already-generated reports based on the date they were generated, and the generate performance report start/end date are the start/end date of the time window which a new report will analyze data from

rgelmis commented:
> > Just a general question on this functionality since I am able to run it now (yay!)- what is the difference between selecting a date under "Issue date" vs generate manual report > generate performance report > start date/end date?
> 
> The issued date is searching for already-generated reports based on the date they were generated, and the generate performance report start/end date are the start/end date of the time window which a new report will analyze data from

Ahhh, I see. I briefly saw a message saying the reports take 10-15 minutes to load and to refresh the page to check back. Is there a way to have a persistent message saying "report is being generated" then update the message when the report is ready, to let the user know to refresh the page? If not, can we keep that initial message displayed for a longer amount of time?

jacob6838 committed:
updating report dialogs to clarify generation timing

jacob6838 commented:
> > > Just a general question on this functionality since I am able to run it now (yay!)- what is the difference between selecting a date under "Issue date" vs generate manual report > generate performance report > start date/end date?
> > 
> > 
> > The issued date is searching for already-generated reports based on the date they were generated, and the generate performance report start/end date are the start/end date of the time window which a new report will analyze data from
> 
> Ahhh, I see. I briefly saw a message saying the reports take 10-15 minutes to load and to refresh the page to check back. Is there a way to have a persistent message saying "report is being generated" then update the message when the report is ready, to let the user know to refresh the page? If not, can we keep that initial message displayed for a longer amount of time?

There was some confusion on our side... Now, the toast notification displays the "this may take up to 15 minutes" message while generating the report, and then refreshes the report list after the report is successfully generated. For our very limited quantity of sample data, this finishes pretty much instantly

To summarize the user experience, the user would press generate report, then a pending toast notification with a spinning loading ring (no percentage feedback currently) would read "Generating Performance Report - may take up to 15 minutes (started at 11:57:02 AM)". Then when the report generation is finished, the notification will turn into a green success toast, and the page will pull for the latest list of reports

rgelmis reviewed (approved):
lgtm! ✨
